numpyro/util.py

```diff
@@ -104,3 +109,73 @@ def scan(f, a, bs):
         return tree_unflatten(tree_def, as_)
     else:
         return lax.scan(f, a, bs)
+
+
+def tscan(f, a, bs, fields=(0,)):
+    """
+    Works as jax.lax.scan but has additional `fields` arguments to select only
+    necessary fields from `a`'s structure. Defaults to selecting only the first
+    field. Other fields will be filled by None.
```

Comment:
Hmm..while putting in None is needed to have a sane function output and signature (e.g. you can't construct a partially filled namedtuple), I wonder if this can be more elegantly handled by specifying a transformation function between whatever pytree data structure we are using and an output data structure which can provided to `tscan`.

Reply:
That would be nice if we can do that but I can't figure out how to implement it for general pytree. I imagine given a transformation, we have to get `fields` (which are indices of required fields for output), then from the output of fori_loop, we convert it to the form we need.